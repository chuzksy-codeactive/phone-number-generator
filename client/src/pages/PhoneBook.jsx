/* eslint-disable react/no-did-update-set-state */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable require-jsdoc */
/* eslint-disable valid-jsdoc */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import phoneBook from '../actions/phoneBookActions';

import './PhoneBook.scss';

/**
 * This is the component for PhoneBood
 */
export class PhoneBook extends Component {
  static defaultProps = {
    phoneBook: {},
  };

  state = {
    isFetching: true,
    phoneBook: {},
    isASC: false,
    isDESC: false,
    amount: '',
    error: false,
  };

  async componentDidMount() {
    await this.props.generatePhoneNumbers(2);
    this.setState(() => ({ isFetching: false, phoneBook: this.props.phoneBook }));
  }

  componentDidUpdate(prevProps) {
    if (JSON.stringify(prevProps.phoneBook) !== JSON.stringify(this.props.phoneBook)) {
      this.setState(() => ({ phoneBook: this.props.phoneBook }));
    }
  }

  handleSordOrder = (sort, phoneNumbers) => {
    let result = [];
    let isASC = false;
    let isDESC = false;

    if (sort === 'ASC') {
      result = phoneNumbers.sort((a, b) => a - b);
      isASC = true;
      isDESC = false;
    } else {
      result = phoneNumbers.sort((a, b) => b - a);
      isASC = false;
      isDESC = true;
    }

    this.setState(prevState => ({
      phoneBook: {
        ...prevState.phoneBook,
        phoneNumbers: result,
      },
      isASC,
      isDESC
    }));
  }

  handleAmountToDisplay = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
      error: false,
    });
  };

  handleGenerateNumber = async () => {
    const amount = Number.parseInt(this.state.amount, 10);

    if (Number.isNaN(amount)) return this.setState(() => ({ error: true }));
    await this.props.generatePhoneNumbers(amount);
  };

  handleGetAllPhoneNumbers = () => {
    this.props.getAllPhoneNumbers();
  }

  handleClearAllPhoneNumbers = () => {
    this.props.clearAllPhoneNumbers();
  }

  header = () => (
    <div className="header">
      <p>RANDOM PHONE NUMBER GENERATOR</p>
    </div>
  );

  dashboard = ({ min, max, count }) => (
    <div className="dashboard">
      {this.card(count, 'Total generated numbers')}
      {this.card(min, 'Minimum phone number')}
      {this.card(max, 'Maximum phone number')}
    </div>
  );

  card = (value, str) => (
    <div className="card">
      <p>{value}</p>
      <p>{str}</p>
    </div>
  );

  phoneList = phoneNumbers => (
    <div className="list">
      <ul className="phone-list">
        {phoneNumbers && phoneNumbers.map((phoneNumber, i) => <li key={phoneNumber}>{phoneNumber}</li>)}
      </ul>
    </div>
  )

  inputControls = numbers => (
    <div className="input-controls">
      <div>Sort Phone Number</div>
      <div className="sort">
        <button
          type="button"
          className={this.state.isASC ? 'order' : 'not-order'}
          onClick={() => this.handleSordOrder('ASC', numbers)}
        >
          Ascending
        </button>
        <button
          type="button"
          className={this.state.isDESC ? 'order' : 'no-order'}
          onClick={() => this.handleSordOrder('DESC', numbers)}
        >
          Descending
        </button>
      </div>
      <div>Enter amount to generate</div>
      {this.state.error && <span>Pls, provide a digit here!</span>}
      <input
        className={this.state.error ? 'has-error' : 'no-error'}
        name="amount"
        type="text"
        value={this.state.amount}
        onChange={this.handleAmountToDisplay}
      />
      <button
        type="button"
        className="btn"
        onClick={this.handleGenerateNumber}
      >
        Generate Phone Numbers
      </button>
      <button
        type="button"
        className="btn"
        onClick={this.handleGetAllPhoneNumbers}
      >
        Get All Phone Numbers
      </button>
      <button
        type="button"
        className="btn-delete"
        onClick={this.handleClearAllPhoneNumbers}
      >
        Clear All Phone Numbers
      </button>
    </div>
  );

  /**
   * renders the phone book compoent
   *
   * @return {JSX} render
   */
  render() {
    const {
      message,
      data: {
        phoneNumbers,
        getMinMaxPhoneNumbers: { min, max } = {}
      } = {}
    } = this.state.phoneBook;
    const { isFetching } = this.state;
    return (
      <div className="container">
        {!isFetching && this.header()}
        {!isFetching && this.dashboard({ min, max, count: phoneNumbers.length })}
        {!isFetching && this.phoneList(phoneNumbers)}
        {!isFetching && this.inputControls(phoneNumbers)}
      </div>
    );
  }
}

PhoneBook.propTypes = {
  generatePhoneNumbers: PropTypes.func.isRequired,
  getAllPhoneNumbers: PropTypes.func.isRequired,
  clearAllPhoneNumbers: PropTypes.func.isRequired,
  phoneBook: PropTypes.exact({
    message: PropTypes.string,
    data: PropTypes.exact({
      phoneNumbers: PropTypes.array.isRequired,
      getMinMaxPhoneNumbers: PropTypes.exact({
        min: PropTypes.string.isRequired,
        max: PropTypes.string.isRequired,
      }),
      message: PropTypes.string,
    })
  }),
};

export const mapStateToProps = state => ({
  ...state,
});

export const mapDispatchToProps = dispatch => ({
  generatePhoneNumbers: (limit, sort) => dispatch(phoneBook.getPhoneNumbers(limit, sort)),
  getAllPhoneNumbers: () => dispatch(phoneBook.getSavePhoneNumbers()),
  clearAllPhoneNumbers: () => dispatch(phoneBook.clearFileContent()),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneBook);
