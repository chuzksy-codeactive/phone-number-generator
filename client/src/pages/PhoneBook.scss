.container {
  @import url('https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700');
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-column-gap: 15px;
  grid-row-gap: 15px;
  font-family: 'Open Sans', sans-serif;

  .header {
    display: flex;
    align-items: center;
    padding: 20px 50px;
    margin-bottom: 20px;
    height: 50px;
    background-color: #0E4D92;
    grid-column: 1 / span 3;
    -moz-box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
    -webkit-box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
    box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);

    p {
      font-size: 1.5rem;
      color: white;
    }
  }

  .dashboard {
    display: flex;
    grid-column: 1 / span 3;
    padding-left: 50px;

    .card:nth-child(1) p:nth-child(1) {
      color: red;
    }

    .card:nth-child(2) p:nth-child(1) {
      color: blue;
    }

    .card:nth-child(3) p:nth-child(1) {
      color: green;
    }

    .card {
      margin-right: 40px;
      

      p:first-child {
        font-size: 50px;
        font-weight: 600;
      }

      p:last-child {
        color: #555;
        font-size: 14px;
        font-weight: 600;
      }
    }
  }

  .list {
    grid-column: 1 / span 2;
    background-color: white;
    margin-top: 50px;
    margin-left: 50px;
    padding: 30px;
    -moz-box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);
    -webkit-box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);

    .phone-list {
      display: grid;
      grid-template-columns: repeat(6, 1fr);
      grid-column-gap: 20px;
      grid-row-gap: 10px;
      align-content: start;
      list-style: none;
      height: 400px;
      overflow: hidden !important;
      overflow-y: scroll !important;
      padding: 0;
      margin: 0;

      li {
        padding: 10px 0;
        border: 1px solid #555;
        background-color: #0E4D92;
        color: white;
        text-align: center;
        font-weight: 300;
        border-radius: 3px;
        font-size: 14px;
      }
    }
  }

  .input-controls {
    display: flex;
    flex-direction: column;
    background-color: white;
    height: 400px;
    margin-right: 50px;
    margin-left: 20px;
    margin-top: 50px;
    padding: 30px;
    -moz-box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);
    -webkit-box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.1);

    input {
      height: 30px;
      margin-top: 10px;
      padding: 0 10px;
      border: 1px solid #0E4D92;
    }

    .btn {
      margin-top: 10px;
      width: 100%;
      height: 30px;
      font-size: 13px;
      border: 1px solid #0E4D92;
    }

    .btn-delete {
      margin-top: 40px;
      background-color: red;
      color: white;
      height: 30px;
      border: 1px solid #0E4D92;
      font-size: 13px;
    }
  }

  .no-phone-book {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
  }
}

p {
  padding: 0;
  margin: 0;
}

.sort {
  display: flex;
  margin-bottom: 60px;
  button {
    margin-top: 10px;
    width: 30%;
    height: 30px;
    font-size: 13px;
    border: 1px solid #0E4D92;
  }
}
  button:hover {
    background-color: #0E4D92;
    color: white;
  }

body {
  margin: 0;
  padding: 0;
  background-color: #F5F5DC;
}

.order {
  background-color: #0E4D92;
  color: white;
}

.no-order {
  background-color: white;
  color:#0E4D92;
}

.has-error {
  border: 1px solid red !important;
}

.no-error {
  border: none;
}

span {
  font-size: 10px;
  font-weight: 300;
  color: red;
  margin-top: 5px;
}