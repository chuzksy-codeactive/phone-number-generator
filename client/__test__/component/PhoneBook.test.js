import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { PhoneBook, mapStateToProps, mapDispatchToProps } from '../../src/pages/PhoneBook';

Enzyme.configure({ adapter: new Adapter() });

describe('Generate phone number', () => {
  const response = {
    data: {
      phoneNumber: [],
      getMinMaxPhoneNumbers: {
        min: '0',
        max: '0'
      }
    },
    message: ''
  };
  const props = {
    generatePhoneNumbers: jest.fn(() => Promise.resolve()),
    getAllPhoneNumbers: jest.fn(() => Promise.resolve()),
    clearAllPhoneNumbers: jest.fn(),
    phoneBook: {
      data: {
        phoneNumbers: ['0808331345', '0154646646', '0254646646'],
        getMinMaxPhoneNumbers: {
          min: '0808331345',
          max: '0154646646'
        }
      }
    }
  };

  const wrapper = shallow(<PhoneBook {...props} />);

  it('should render the <PhoneBook /> component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should sort the phone-numbers in ascending order', () => {
    const handleSordOrder = jest.fn();

    wrapper.find('.sort button').first().simulate('click');
    wrapper.instance().handleSordOrder('ASC', props.phoneBook.data.phoneNumbers);

    expect(wrapper.state().phoneBook.data.phoneNumbers).toStrictEqual(['0154646646', '0254646646', '0808331345']);
  });

  it('should sort the phone-numbers in descending order', () => {
    const handleSordOrder = jest.fn();

    wrapper.find('.sort button').last().simulate('click');
    wrapper.instance().handleSordOrder('DESC', props.phoneBook.data.phoneNumbers);

    expect(wrapper.state().phoneBook.data.phoneNumbers).toStrictEqual(['0808331345', '0254646646', '0154646646']);
  });

  it('should call "handleGenerateNumber()" on button click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleGenerateNumber');
    const event = { target: { name: 'amount', value: '10' } };

    wrapper.update();
    wrapper.find('input[name="amount"]').simulate('change', event);
    wrapper.find('.btn').first().simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it('should throw error when user does not pass a digit', () => {
    const event = { target: { name: 'amount', value: 'eye' } };

    wrapper.update();
    wrapper.find('input[name="amount"]').simulate('change', event);
    wrapper.find('.btn').first().simulate('click');
    wrapper.instance().handleGenerateNumber();

    expect(wrapper.state().error).toBe(true);
  });

  it('should call "handleClearAllPhoneNumbers" on button click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleClearAllPhoneNumbers');
    const event = { target: { name: 'amount', value: '10' } };

    wrapper.update();
    wrapper.find('input[name="amount"]').simulate('change', event);
    wrapper.find('.btn-delete').simulate('click');

    expect(spy).toHaveBeenCalled();
  });

  it('should call "handleGetAllPhoneNumbers" on button click', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleGetAllPhoneNumbers');
    const event = { target: { name: 'amount', value: '10' } };

    wrapper.update();
    wrapper.find('input[name="amount"]').simulate('change', event);
    wrapper.find('.btn').last().simulate('click');

    expect(spy).toHaveBeenCalled();
  });
});

describe('mapStateToProps', () => {
  it('returns the expected props object', () => {
    const state = {};

    const props = mapStateToProps(state);
    expect(props.phoneBook).toEqual(state.phoneBook);
  });
});

describe('mapDispatchToProps', () => {
  let dispatch;
  let props;

  beforeEach(() => {
    dispatch = jest.fn();
    props = mapDispatchToProps(dispatch);
  });

  afterEach(() => {
    dispatch = props = null;
  });

  it('ensures that generatePhoneNumbers is mapped to props', () => {
    props.generatePhoneNumbers();
    expect(dispatch).toBeCalled();
  });

  it('ensures that getAllPhoneNumbers is mapped to props', () => {
    props.getAllPhoneNumbers();
    expect(dispatch).toBeCalled();
  });

  it('ensure clearAllPhoneNumbers is mapped to props', () => {
    props.clearAllPhoneNumbers();
    expect(dispatch).toHaveBeenCalled();
  });
});
