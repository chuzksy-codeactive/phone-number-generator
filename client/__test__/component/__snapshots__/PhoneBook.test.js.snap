// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate phone number should render the <PhoneBook /> component correctly 1`] = `ShallowWrapper {}`;
